src/cocaine-app/sched/__init__.py

```diff
@@ -0,0 +1,417 @@
+        for res_type in jobs.Job.RESOURCE_TYPES:
+            for job_type in jobs.JobTypes.AVAILABLE_TYPES:
+                self.res_limits[job_type][res_type] = config.get('jobs', {}).get(job_type, {}).get(
+                            'resources_limits', {}).get(res_type, 1)
+        logger.info("Obtained res limits are {}".format(self.res_limits))
+
+    def _start_tq(self):
+        self.__tq.start()
+
+    def register_periodic_func(self, starter_func, period_default, starter_name=None, lock_name=None):
+        """
+        Guarantee that starter_func would be run within specified period under a zookeeper lock
+        :param starter_func - a function to be run
+        :param period_default - the default value of period to run the function with
+            if the "{starter_name}_period" is not specified under {starter_name} sub-section with sched config section
+        :param starter_name - a logical name of starter. Name of starter subsection within sched section in config.
+                            Defaults to starter_func.__name__
+        :param lock_name - a name of lock in zookeeper that protects starter from simultaneous execution).
+                            Defaults to "scheduler/{starter_name}
+        """
+
+        starter_name = starter_name or starter_func.__name__
+        lock_name = lock_name or "scheduler/{}".format(starter_name)
+        period_param = "{}_period".format(starter_name)
+        period_val = self.params.get(starter_name, {}).get(period_param, period_default)
+
+        logger.info("Registering periodic starter func {} to be raised on {} period".format(starter_name, period_val))
+
+        if not self.params.get(starter_name, {}).get('enabled', False):
+            logger.info("Starter {} is disabled".format(starter_name))
+            return
+
+        timer = periodic_timer(seconds=period_val)
+
+        def _starter_periodic_func():
+            try:
+                logger.info('Starting {}'.format(starter_name))
+
+                with sync_manager.lock(lock_name, blocking=False):
+                    starter_func()
+
+            except LockFailedError:
+                logger.info('Another {} is already running under {} lock'.format(starter_func.__name__, lock_name))
+            except:
+                logger.exception('Failed to {}'.format(starter_name))
+            finally:
+                logger.info('{} finished'.format(starter_name))
+                self.__tq.add_task_at(
+                    starter_name,
+                    timer.next(),
+                    _starter_periodic_func)
+
+        self.__tq.add_task_at(
+            starter_name,
+            timer.next(),
+            _starter_periodic_func
+        )
+
+        starter = {
+            "starter_func": starter_func,
+            "starter_name": starter_name,
+            "starter_lock": lock_name,
+            "starter_timer": timer,
+        }
+
+        self.starters.append(starter)
```

Comment:
Fixed in https://github.com/yandex/mastermind/pull/417/commits/69574909fafa454bb1f62443cd1b2b78de089e76